stdlib/path.gr

```diff
@@ -0,0 +1,571 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+enum Token {
+  Slash,
+  Dot,
+  Dotdot,
+  DriveTok(Char),
+  Text(String),
+}
+
+/**
+ * @section Types: Type declarations included in the Path module.
+ */
+
+/**
+ * Represents an absolute path's anchor point.
+ */
+export enum AbsoluteRoot {
+  Root,
+  Drive(Char),
+}
+
+// Rel Number represents number of directories up from base point
+enum BasePath<a> {
+  Rel(Number),
+  Abs(AbsoluteRoot),
+}
+
+enum FileType<a> {
+  File,
+  Dir,
+}
+
+// NOTE dummy record names put here just to distinguish the two. These could be
+// replaced with opaque types if they get added to the language
+/**
+ * Opaque type representing a relative path.
+ */
+record Relative {
+  _rel: Void,
+}
+
+/**
+ * Opaque type representing an absolute path.
+ */
+record Absolute {
+  _abs: Void,
+}
+
+/**
+ * Opaque type representing an path referencing a file.
+ */
+record File {
+  _file: Void,
+}
+
+/**
+ * Opaque type representing an path referencing a directory.
+ */
+record Directory {
+  _directory: Void,
+}
+
+/**
+ * Opaque type representing a system path, parameterized on
+ * (relative|absolute, file|directory).
+ */
+type Path<a, b> = (BasePath<a>, FileType<b>, List<String>)
+
+/**
+ * Wraps each possibile path type to a variant; can be used to dynamically
+ * determine the type of the path wrapped in cases where it is needed. It is
+ * recommended to only make use of this type in cases where it is unavoidable.
+ */
+export enum PathWithType {
+  AbsoluteFile(Path<Absolute, File>),
+  AbsoluteDir(Path<Absolute, Directory>),
+  RelativeFile(Path<Relative, File>),
+  RelativeDir(Path<Relative, Directory>),
+}
+
+/**
+ * Represents a platform-specific path encoding scheme.
+ */
+export enum PathEncoding {
+  Windows,
+  Posix,
+}
+
+/**
+ * Represents possible errors for the `relativeTo` operation.
+ */
+export enum RelativizationError {
+  DifferentRoots,
+  ImpossibleRelativization,
+}
+
+/**
+ * @section Values: Functions for working with Paths.
+ */
+
+let makeToken = str => {
+  let len = String.length(str)
+  match (str) {
+    "." => Dot,
+    ".." => Dotdot,
+    _ when len == 2 && String.charAt(1, str) == ':' =>
+      DriveTok(String.charAt(0, str)),
+    _ => Text(str),
+  }
+}
+
+let rec parseNextToken = (path: Path<a, b>, nextToken) => {
+  let (base, ft, subs) = path
+  match (nextToken) {
+    Slash | Dot => path,
+    DriveTok(label) => (base, ft, [Char.toString(label) ++ ":", ...subs]),
+    Text(str) => (base, ft, [str, ...subs]),
+    Dotdot => {
+      match (path) {
+        (_, _, [_, ...rest]) => (base, ft, rest),
+        (Rel(upDirs), _, []) => (Rel(upDirs + 1), ft, []),
+        (Abs(_), _, []) => path,
+      }
+    },
+  }
+}
+
+// splits a path on forward slashes
+let lexPath = pathStr => {
+  let pathStr = String.trim(pathStr)
```

Comment:
Hmm, in hindsight maybe not actually; I initially thought that the main OSes disallow leading and trailing spaces in file names but it turns out that's not always the case so it might be best to remove the trim